File: csharp/1399.count-largest-group.cs
/*
 * @lc app=leetcode id=1399 lang=csharp
 *
 * [1399] Count Largest Group
 */

using System.Linq;

// @lc code=start
public class Solution {
    public int CountLargestGroup(int n) {
        var grouped = Enumerable
            .Range(1, n)
            .GroupBy(x => {
                int ret = 0;
                while (x > 0) {
                    ret += x % 10;
                    x /= 10;
                }
                return ret;
            }, (x, a) => a.Count())
            .GroupBy(t => t, (t, a) => (Value: t, Count: a.Count()))
            .ToDictionary(t => t.Value, t => t.Count);
        return grouped[grouped.Keys.Max()];
    }
}
// @lc code=end

File: java/1399.count-largest-group.java
/*
 * @lc app=leetcode id=1399 lang=java
 *
 * [1399] Count Largest Group
 */

// @lc code=start
class Solution {
    public int countLargestGroup(int n) {
        int k = 1, t = n;
        while (t > 0) {
            k += 9;
            t /= 10;
        }
        int[] count = new int[k];
        for (int i = 1; i <= n; ++i) {
            t = i;
            int sum = 0;
            while (t > 0) {
                sum += t % 10;
                t /= 10;
            }
            ++count[sum];
        }
        int ret = -1, max = -1;
        for (int i = 1; i < k; ++i) {
            if (count[i] == max) {
                ++ret;
            } else if (count[i] > max) {
                max = count[i];
                ret = 1;
            }
        }
        return ret;
    }
}
// @lc code=end

